### PR TITLE
Switch to Hanami::Model for repositories/entities

### PR DESCRIPTION
**Pull request by mellowfish**
`Hanami::Model` is a more formal implementation of the Repository pattern. You instantiate a repository, call a method on it, and get back immutable structs instead of mutable record instances.

I had to do a small amount of refactoring because of the immutable nature of the entities, but it actually pushed me in a better direction, as an `AuthenticatedUser` decorator is a really nice idiom.

See https://github.com/hanami/model for more details.

### CODE REVIEW

File: app/repositories/user_repository.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+class UserRepository < Hanami::Repository
+  def sign_up(session: nil, **create_params)
```
Comment (jwsloan) [inline app/repositories/user_repository.rb:2]: Do `sign_up` and `authenticate` really belong in the Repository? I thought it was meant just for data retrieval, not actions like those. I would expect something along the lines of `User::SignUp.call` and `User::Authenticate.call`. Or maybe `Authentication.new.sign_up`.

Comment (mellowfish) [reply app/repositories/user_repository.rb:2]: So, I *have* a `User::Create` and `Session::Create`. But I wanted all the authentication (password hashing and checking) done in one place. I could see if I could extract it to another dedicated class//module that the others rely on.

**Review by brooksquil — APPROVED** [review]
This does seem like a cleaner way of doing this, got me to do a lot of learning about repository pattern. Thank you.

File: app/repositories/user_repository.rb (hunk 1 above)
Comment (jwsloan) [reply app/repositories/user_repository.rb:2]: I think some additional layer of abstraction is warranted.

Comment (mellowfish) [reply app/repositories/user_repository.rb:2]: I'll take a look at it this weekend.

**Review by jwsloan — APPROVED** [review]
Let's go ahead and get this merged. We can look into refactoring the sign up stuff in a separate PR.

**Pull request merged**